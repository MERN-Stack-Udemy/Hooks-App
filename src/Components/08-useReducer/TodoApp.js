import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer'
import { useForm } from '../../Hooks/useForm';

import './styles.css';

export const TodoApp = () => {

  const [ { description }, handleChage, reset ] = useForm({
    description:''
  });

  const init = () => {
    return JSON.parse( localStorage.getItem('todos') ) || [];
  }

  const [ todos, dispatch ] = useReducer(todoReducer, [], init);
  
  useEffect(() => {
    localStorage.setItem( 'todos', JSON.stringify(todos) );
  }, [todos])

  const handleSubmit = e => { 
    e.preventDefault();

    if( description.trim().length <=2){
      return;
    }
    
    const newToDo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    }

    const action = {
      type: 'ADD',
      payload: newToDo
    }
    
    dispatch( action );
    reset();
  }

  const handleToggle = ( todoId ) => {
    dispatch({
      type: 'TOGGLE',
      payload: todoId
    });
  }

  const handleDelete = todoId => {
    const action = {
      type: 'DELETE',
      payload: todoId
    };
    dispatch( action );
  } 

  return (
    <div>
      <h1>Todos ({todos.length})</h1>
      <hr/>

      <div className="row">
        <div className="col-7">
          <ul className="list-group list-group-flush">
            {
              todos.map((todo, i) => (
                <li 
                  key={todo.id}
                  className="list-group-item"
                >
                  <p
                    className={ `${ todo.done && 'complete'}`}
                    onClick={()=> handleToggle( todo.id )}
                  >{i+1}. {todo.desc}</p>
                  <button 
                    className="btn btn-danger"
                    onClick= { () => handleDelete(todo.id) }
                  >Delete</button>
                </li>
              ))
            }
          </ul>
        </div>
        <div className="col-5">

          <h5>Add ToDo</h5>
          <hr/>
          <form onSubmit={ handleSubmit }>
            <input
              autoComplete="off"
              className="form-control"
              name="description"
              onChange={ handleChage }
              placeholder="To Do..."
              type="text"
              value={ description }
            />
            
            <button
              className="btn btn-outline-primary mt-2 btn-block"
              type="submit"
            >
              Add
            </button>

          </form>

        </div>
      </div>


    </div>
  )
}
