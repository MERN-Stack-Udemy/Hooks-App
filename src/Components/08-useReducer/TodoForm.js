import React from "react";
import { useForm } from "../../Hooks/useForm";

export const TodoForm = ({ dispatch }) => {

  const [ { description }, handleChage, reset ] = useForm({
    description:''
  });

  const handleSubmit = e => { 
    e.preventDefault();

    if( description.trim().length <=2){
      return;
    }
    const newToDo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    }

    dispatch({
      type: 'ADD',
      payload: newToDo
    });
    reset();
  }
  
  return (
    <form onSubmit={handleSubmit}>
      <input
        autoComplete="off"
        className="form-control"
        name="description"
        onChange={handleChage}
        placeholder="To Do..."
        type="text"
        value={description}
      />

      <button className="btn btn-outline-primary mt-2 btn-block" type="submit">
        Add
      </button>
    </form>
  );
};
