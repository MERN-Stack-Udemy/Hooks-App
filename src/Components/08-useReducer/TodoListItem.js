import React from 'react'

export const TodoListItem = ({ todo, index, dispatch}) => {
  
  const handleToggle = ( todoId ) => {
    dispatch({
      type: 'TOGGLE',
      payload: todoId
    });
  }

  const handleDelete = todoId => {
    const action = {
      type: 'DELETE',
      payload: todoId
    };
    dispatch( action );
  } 

  return (
    <li
      className='list-group-item'
    >
      <p
        className={ `${ todo.done && 'complete'}`}
        onClick={()=> handleToggle( todo.id )}
      >{index+1}. {todo.desc}</p>
      <button 
        className="btn btn-danger"
        onClick= { () => handleDelete(todo.id) }
      >Delete</button>
    </li>
  )
}
