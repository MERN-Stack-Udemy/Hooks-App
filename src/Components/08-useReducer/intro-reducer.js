
const initialState = [{
  id: 1,
  todo: 'ir a correr',
  done: false
}];

const todoReducer = (state=initialState, action ) => {
  if(action?.type === 'ADD'){
    return [...initialState, action.payload ];
  }

  return state;
}

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: 'estudiar react',
  done: false
}

const addNewAction = {
  type: 'ADD',
  payload: newTodo,
  // dispatch
}

todos = todoReducer( todos, addNewAction );


console.log(todos); 