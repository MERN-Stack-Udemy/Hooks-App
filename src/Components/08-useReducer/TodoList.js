import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, dispatch}) => {
  return (
    <ul className="list-group list-group-flush">
      {
        todos.map( (todo, index) => (
          <TodoListItem 
            dispatch={dispatch}
            index={index}
            key = {todo.id}
            todo={todo}
          />
        ) )
      }
    </ul>
  )
}
