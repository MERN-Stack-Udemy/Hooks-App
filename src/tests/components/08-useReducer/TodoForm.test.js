import {shallow} from 'enzyme';
import { TodoForm } from '../../../Components/08-useReducer/TodoForm';

describe("test on Todo Form add todo", () => {
  
  const dispatch = jest.fn();

  const wrapper = shallow(
    <TodoForm
      dispatch = {dispatch}
    />
  );

  test("should be display correctly", () => {
    expect( wrapper ).toMatchSnapshot();
  });

  test('not should call \'dispatch\' function', () => {

      const formSubmit = wrapper.find('form').prop('onSubmit');
      formSubmit({ preventDefault(){} });
      expect( dispatch ).toHaveBeenCalledTimes(0);

  })

  test('should call \'handleSubmit\' with one argument', () => {
    
    const value = 'learn React';
    wrapper.find('input').simulate('change',{
      target:{
        value,
        name: 'description'
      }
    });

    const formSubmit = wrapper.find('form').prop('onSubmit');
      formSubmit({ preventDefault(){} });
      expect( dispatch ).toHaveBeenCalledTimes(1);
      expect( dispatch ).toBeCalledWith( expect.any(Object) );
      expect( dispatch ).toBeCalledWith({
        type : 'ADD',
        payload :{
          id : expect.any(Number),
          desc : value,
          done : false
        }
      });
      // when submit is called, also call 'useForm.reset()', then the value of input shold it's empty
      expect( wrapper.find('input').prop('value') ).toBe('');
  })
  
  
});
