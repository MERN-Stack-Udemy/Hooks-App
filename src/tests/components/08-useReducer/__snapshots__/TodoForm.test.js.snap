// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on Todo Form add todo should be display correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    className="form-control"
    name="description"
    onChange={[Function]}
    placeholder="To Do..."
    type="text"
    value=""
  />
  <button
    className="btn btn-outline-primary mt-2 btn-block"
    type="submit"
  >
    Add
  </button>
</form>
`;
